pageserver/src/storage_sync/download.rs

```diff
@@ -233,89 +243,128 @@ pub(super) async fn download_timeline_layers<'a>(
-                //     fsync(tmp)
-                //     rename(tmp, new)
-                //     fsync(new)
-                //     fsync(parent)
-                // For more context about durable_rename check this email from postgres mailing list:
-                // https://www.postgresql.org/message-id/56583BDD.9060302@2ndquadrant.com
-                // If pageserver crashes the temp file will be deleted on startup and re-downloaded.
-                let temp_file_path =
-                    path_with_suffix_extension(&layer_destination_path, TEMP_FILE_SUFFIX);
-
-                let mut destination_file =
-                    fs::File::create(&temp_file_path).await.with_context(|| {
-                        format!(
-                            "Failed to create a destination file for layer '{}'",
-                            temp_file_path.display()
-                        )
-                    })?;
-
-                let mut layer_download = storage.download_storage_object(None, &layer_destination_path)
-                    .await
-                    .with_context(|| {
-                        format!(
-                            "Failed to initiate the download the layer for {sync_id} into file '{}'",
-                            temp_file_path.display()
-                        )
-                    })?;
-                io::copy(&mut layer_download.download_stream, &mut destination_file)
-                    .await
-                    .with_context(|| {
-                        format!(
-                            "Failed to download the layer for {sync_id} into file '{}'",
-                            temp_file_path.display()
-                        )
-                    })?;
-
-                // Tokio doc here: https://docs.rs/tokio/1.17.0/tokio/fs/struct.File.html states that:
-                // A file will not be closed immediately when it goes out of scope if there are any IO operations
-                // that have not yet completed. To ensure that a file is closed immediately when it is dropped,
-                // you should call flush before dropping it.
-                //
-                // From the tokio code I see that it waits for pending operations to complete. There shouldn't be any because
-                // we assume that `destination_file` file is fully written. I.e there is no pending .write(...).await operations.
-                // But for additional safety let's check/wait for any pending operations.
-                destination_file.flush().await.with_context(|| {
+        .map(|(layer_destination_path, metadata)| async move {
+
+            match layer_destination_path.metadata() {
+                Ok(m) if m.is_file() => {
+                    // the file exists from earlier round when we failed after renaming it as
+                    // layer_destination_path
+                    let verified = if let Some(expected) = metadata.file_size() {
+                        m.len() == expected
```

Comment:
Just to check my understanding: only way this file exists here is by completing this particular download in an earlier attempt with any of the following:
- failing to fsync the directory in earlier attempts
- failing with some other download

The download process skips files (`LayersDownload::from_skipped_layers`) which were present when the download got started (subject to startup time metadata gathering in `collect_timeline_files` and verification in `compare_local_and_remote_timeline`). As far as I know during the download there cannot be any layer creation because of `RemoteTimeline.awaits_download`.

Reply:
> only way this file exists here is by completing this particular download

Not the only way, as demonstrated by the added test case which downloads a truncated layer file from remote on startup: https://github.com/neondatabase/neon/blob/271182a8e769e23b6c427e7952eb41c0db418d26/test_runner/regress/test_tenants_with_remote_storage.py#L330-L392

This test case is rather artificial, but it is in the realm of "things that can happen with filesystems."

I'll add warn logging.